parity-clib/examples/cpp/main.cpp

```diff
@@ -14,172 +14,134 @@
 };
 
 // callback that gets invoked upon an event
-void callback(void* user_data, const char* response, size_t _len) {
-	Callback* cb = static_cast<Callback*>(user_data);
-	if (cb->type == CALLBACK_RPC) {
-		cb->counter -= 1;
-	} else if (cb->type == CALLBACK_WS) {
-		std::regex is_subscription ("\\{\"jsonrpc\":\"2.0\",\"result\":\"0[xX][a-fA-F0-9]{16}\",\"id\":1\\}");
-		if (std::regex_match(response, is_subscription) == true) {
-			cb->counter -= 1;
-		}
-	}
+void callback(std::string_view buf) { (void)buf; }
+} // namespace
+
+int main(int _argc, char **_argv) {
+  (void)_argc, (void)_argv;
+  using parity::ethereum::ParityEthereum;
+  // run full-client
+  if (0) {
+    std::vector<std::string> cli_args{"--no-ipc"s, "--jsonrpc-apis=all"s,
+                                      "--chain"s, "kovan"s};
+    ParityEthereum parity = parity_run(cli_args);
+    parity_rpc_queries(parity);
+    parity_subscribe_to_websocket(parity);
+  }
+
+  // run light-client
+  {
+    std::vector<std::string> light_config = {
+        "--no-ipc"s, "--light"s, "--jsonrpc-apis=all"s, "--chain"s, "kovan"s};
+    ParityEthereum parity = parity_run(light_config);
+    parity_rpc_queries(parity);
+    parity_subscribe_to_websocket(parity);
+    exit(1);
+  }
+  return 0;
 }
 
-int main() {
-	// run full-client
-	{
-		std::vector<const char*> config = {"--no-ipc" , "--jsonrpc-apis=all", "--chain", "kovan"};
-		void* parity = parity_run(config);
-		if (parity_rpc_queries(parity)) {
-			printf("rpc_queries failed\r\n");
-			return 1;
-		}
-
-		if (parity_subscribe_to_websocket(parity)) {
-			printf("ws_queries failed\r\n");
-			return 1;
-		}
-
-		if (parity != nullptr) {
-			parity_destroy(parity);
-		}
-	}
-
-	// run light-client
-	{
-		std::vector<const char*> light_config = {"--no-ipc", "--light", "--jsonrpc-apis=all", "--chain", "kovan"};
-		void* parity = parity_run(light_config);
-
-		if (parity_rpc_queries(parity)) {
-			printf("rpc_queries failed\r\n");
-			return 1;
-		}
-
-		if (parity_subscribe_to_websocket(parity)) {
-			printf("ws_queries failed\r\n");
-			return 1;
-		}
-
-		if (parity != nullptr) {
-			parity_destroy(parity);
-		}
-	}
-	return 0;
+namespace {
+void parity_rpc_queries(ParityEthereum &parity) {
+  Callback cb{.type = CALLBACK_RPC, .counter = rpc_queries.size()};
+  auto cb_func = std::function(cb);
+
+  try {
+    for (const auto &query : rpc_queries)
+      parity.rpc(query, TIMEOUT_ONE_MIN_AS_MILLIS, cb_func);
+  } catch (std::exception &exn) {
+    std::cerr << exn.what() << std::endl;
+    while (cb.counter != 0)
+      ;
+    throw;
+  } catch (...) {
+    while (cb.counter != 0)
+      ;
+    throw;
+  }
+  while (cb.counter != 0)
+    ;
 }
 
-int parity_rpc_queries(void* parity) {
-	if (!parity) {
-		return 1;
-	}
+void parity_subscribe_to_websocket(ParityEthereum &parity) {
+  // MUST outlive the std::vector below
+  Callback cb{.type = CALLBACK_WS, .counter = ws_subscriptions.size()};
```

Comment:
As far as I know designated initializers are a C++20 feature and unfortunately it is still 2019. :(
https://blog.therocode.net/2018/11/exploring-cpp-20-designated-initializers